File: solutions/19.remove-nth-node-from-end-of-list/two-pointers.ts
/*
 * @lc app=leetcode.cn id=19 lang=typescript
 *
 * https://leetcode-cn.com/problems/remove-nth-node-from-end-of-list
 */

/**
 * Definition for singly-linked list.
 */
class ListNode {
    val: number;
    next: ListNode | null;
    constructor(val?: number, next?: ListNode | null) {
        this.val = val === undefined ? 0 : val;
        this.next = next === undefined ? null : next;
    }
}

// @lc code=start
function removeNthFromEnd(head: ListNode | null, n: number): ListNode | null {
    const sentry = new ListNode(0, head);
    // finally: left = target.prev
    let left = sentry;
    // finally: right = tail
    let right = sentry;
    for (let i = 0; i < n; ++i) {
        right = right.next as ListNode;
    }
    while (right.next) {
        right = right.next;
        left = left.next as ListNode;
    }
    left.next = (left.next as ListNode).next;
    return sentry.next;
}
// @lc code=end

File: solutions/19.remove-nth-node-from-end-of-list/src/two_pointers.rs
/*
 * @lc app=leetcode.cn id=19 lang=rust
 *
 * https://leetcode.cn/problems/remove-nth-node-from-end-of-list/
 */

pub struct Solution;

#[derive(PartialEq, Eq, Clone, Debug)]
pub struct ListNode {
    pub val: i32,
    pub next: Option<Box<ListNode>>,
}

impl ListNode {
    #[inline]
    fn new(val: i32) -> Self {
        ListNode { next: None, val }
    }
}

// @lc code=start
impl Solution {
    pub fn remove_nth_from_end(head: Option<Box<ListNode>>, n: i32) -> Option<Box<ListNode>> {
        let mut dummy = Box::new(ListNode { val: 0, next: head });
        let mut p_1 = &dummy as *const Box<ListNode>;
        let mut p_2 = &mut dummy as *mut Box<ListNode>;
        unsafe {
            for _ in 0..n {
                if let Some(node) = &(*p_1).next {
                    p_1 = node;
                } else {
                    return dummy.next;
                }
            }
            while let Some(node) = &(*p_1).next {
                p_1 = node;
                p_2 = (*p_2).next.as_mut().unwrap();
            }
            (*p_2).next = (*p_2).next.as_mut().unwrap().next.take();
        }
        dummy.next
    }
}
// @lc code=end
